File: lib/formulas.mjs
export function goatCalc(listDate, listRank, isRanked, baseValue) {
  const currentDate = new Date().getFullYear();
  const dateDiff = currentDate - listDate;
  const dateCutoff = 10;
  let pointCalc = 0;
  if (isRanked) {
    if (dateDiff > 0 && dateDiff <= dateCutoff) {
      if (listRank != 1) {
        pointCalc = baseValue + (1 - dateDiff / dateCutoff) + 1 / listRank;
      } else {
        pointCalc = baseValue + (1 - dateDiff / dateCutoff) + 1;
      }
    }
    if (dateDiff === 0) {
      if (listRank != 1) {
        pointCalc = baseValue + 1 + 1 / listRank;
      } else {
        pointCalc = baseValue + 1 + 1;
      }
    }
    if (dateDiff > dateCutoff) {
      if (listRank != 1) {
        pointCalc = baseValue + (1 / listRank) * (1 / dateDiff);
      } else {
        pointCalc = baseValue + 1 / dateDiff;
      }
    }
  } else {
    if (dateDiff > 0 && dateDiff <= dateCutoff) {
      pointCalc = baseValue + (1 - dateDiff / dateCutoff);
    }
    if (dateDiff === 0) {
      pointCalc = baseValue + 1;
    }
    if (dateDiff > dateCutoff) {
      pointCalc = baseValue + 1 / dateDiff;
    }
  }
  return pointCalc;
}

export function gotyCalc(listRank) {
  const baseValue = 0.9;
  let pointCalc;
  if (listRank != 1 && listRank != "Unranked")
    pointCalc = baseValue + 1 / (listRank * 10);
  else if (listRank === "Unranked") pointCalc = baseValue;
  else pointCalc = 1;
  return pointCalc;
}

export function gotdCalc(listRank) {
  const baseValue = 1;
  let pointCalc;
  if (listRank != 1 && listRank != "Unranked")
    pointCalc = baseValue + 1 / (listRank * 2);
  else if (listRank === "Unranked") pointCalc = baseValue;
  else pointCalc = baseValue;
  return pointCalc;
}

// export function bestOfGenre(listRank) {
//   const baseValue = 1.5;
//   let pointCalc;
//   if (listRank != 1 && listRank != "Unranked")
//     pointCalc = baseValue + 1 / (listRank * 4);
//   else if (listRank === "Unranked") pointCalc = baseValue;
//   else pointCalc = baseValue;
//   return pointCalc;
// }

File: lib/formulas.py
from datetime import datetime
from decimal import *
from typing import Union


def goat_calc(list_date: int, list_rank: Union[int, None], is_ranked: int, base_value: float):
    current_date: int = datetime.now().year
    date_diff: int = current_date - list_date
    date_cutoff: int = 10

    one_dec: Decimal = Decimal(1)
    base_value_dec: Decimal = Decimal(base_value)

    date_points: Decimal = Decimal(date_diff) / Decimal(date_cutoff)
    point_calc: Union[int, Decimal] = 0

    if is_ranked == 1:
        rank_points: Decimal = one_dec / Decimal(list_rank)
        if 0 < date_diff <= date_cutoff:
            if list_rank != 1:
                point_calc = base_value_dec + (one_dec - date_points) + rank_points
            else:
                point_calc = base_value_dec + (Decimal(2) - date_points)

        if date_diff == 0:
            if list_rank != 1:
                point_calc = base_value_dec + one_dec + rank_points
            else:
                point_calc = base_value_dec + Decimal(2)

        if date_diff > date_cutoff:
            if list_rank != 1:
                point_calc = base_value_dec + rank_points * (one_dec / Decimal(date_diff))
            else:
                point_calc = base_value_dec + (one_dec / Decimal(date_diff))

    else:
        if 0 < date_diff <= date_cutoff:
            point_calc = base_value_dec + (one_dec - date_points)
        if date_diff == 0:
            point_calc = base_value_dec + one_dec
        if date_diff > date_cutoff:
            point_calc = base_value_dec + (one_dec / Decimal(date_diff))

    return point_calc


def goty_calc(list_rank):
    base_value: Decimal = Decimal(0.9)

    if list_rank != 1 and list_rank != 'Unranked' and list_rank is not None:
        points: Decimal = Decimal(1) / (Decimal(list_rank) * Decimal(10))
        return base_value + points
    elif list_rank == 'Unranked':
        return base_value
    else:
        return Decimal(1)
